pkg/services/ngalert/state/manager.go

```diff
@@ -153,17 +153,21 @@ func (st *Manager) setNextState(alertRule *ngModels.AlertRule, result eval.Resul
 	currentState.TrimResults(alertRule)
 
 	st.log.Debug("setting alert state", "uid", alertRule.UID)
+	var err error
 	switch result.State {
 	case eval.Normal:
-		currentState.resultNormal(result)
+		err = currentState.resultNormal(alertRule, result)
```

Comment:
I wondering if it is better if the creation of annotations is non-blocking / async to it doesn't interfere with higher priority state processing. Since we are only logging, can drop the error here and move it to annotation creation which I think would make making it async easier.

Reply:
We can. I chose to return the error since the State doesn't have a logger. We could pass the manager's logger as a param but I'm not sure off the top of my head how that would impact synchronicity. IMO the fact that the rule is evaluated asyncrhonously is probably sufficient parallelization.

Reply:
My concern is that the annotations I think are backed by SQL (by default at least), so if there is an additional DB write on each of these that could pile up - but maybe annotations don't work like that?

Reply:
They definitely do. Theoretically it would only back up on transitions. Which can happen to every match at once, easily, but subsequent evaluations would be unlikely to flap back and forth, creating a large backlog of inserts. Famous last words, I know. We could possibly return whether a state change occurred and batch the annotation creation for one call at the end of processing the result set. It would require some changes to the annotation package to add support for saving a batch. Could work though. What do you think?

Reply:
I think since annotations long term have been mentioned to me as something that should change (e.g. make Loki a provider, maybe default provider?) maybe better to stay out of the internals. But could add a BulkSave method to the Repository interface I think.

I guess I still think the async call to write the annotation would be best even it is a bulk operation. If avoiding bulk, they could maybe go onto a capped channel and we could drop them if there are too many queued inserts.

Reply:
@kylebrandt I moved the annotation creation into the manager, which, upon reflection, is really where it belongs anyway. This gives it access to the logger, which I believe is thread safe. I'm just calling it using `go` now. This means that, during shutdown, there might be some unwritten annotations. I think this is acceptable given the possible trade off of significantly increasing the shutdown time, in the event that annotations are slow.